### PR TITLE
Add DisposeAsync support to generic host

### PR DESCRIPTION
**Pull request by pakrym**
aspnet/Extensions#426

### CODE REVIEW

File: src/Hosting/Hosting/test/Internal/HostTests.AsyncDisposable.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#if NETCOREAPP
```
Comment (Tratcher) [inline src/Hosting/Hosting/test/Internal/HostTests.AsyncDisposable.cs]: netcoreapp3.0? Or use the same DISPOSE_ASYNC

**Comment by pakrym** [comment]
> What about the places where Host.Dispose is currently called?

What about them?

File: src/Hosting/Hosting/test/Internal/HostTests.AsyncDisposable.cs (hunk 1 above)
Comment (pakrym) [reply src/Hosting/Hosting/test/Internal/HostTests.AsyncDisposable.cs]: No need to be specific to 3.0, every netcoreapp past 3.0 would work.

**Comment by davidfowl** [comment]
What about RunAsync?

**Comment by pakrym** [comment]
Done

File: src/Hosting/Hosting/src/Internal/Host.cs [hunk 2]
```diff
@@ -98,9 +101,29 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
...
 
+#if DISPOSE_ASYNC
+        public void Dispose()
+        {
+            DisposeAsync().GetAwaiter().GetResult();
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            switch (Services)
```
Comment (Tratcher) [inline src/Hosting/Hosting/src/Internal/Host.cs]: does switch work on null?

File: src/Hosting/Hosting/test/Internal/HostTests.AsyncDisposable.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Xunit;
+
+namespace Microsoft.Extensions.Hosting.Internal
+{
+    public partial class HostTests
+    {
+        [Fact]
+        public async Task HostCallsDisposeAsyncOnServiceProvider()
```
Comment (Tratcher) [inline src/Hosting/Hosting/test/Internal/HostTests.AsyncDisposable.cs]: No test that directly calls DisposeAsync?

File: src/Hosting/Hosting/src/Internal/Host.cs (hunk 2 above)
Comment (pakrym) [reply src/Hosting/Hosting/src/Internal/Host.cs]: Yes

File: src/Hosting/Hosting/test/Internal/HostTests.AsyncDisposable.cs (hunk 3 above)
Comment (pakrym) [reply src/Hosting/Hosting/test/Internal/HostTests.AsyncDisposable.cs]: Added

**Pull request merged**